packages/agent/src/client.ts

```diff
@@ -0,0 +1,108 @@
+import { fork, monitor, Operation } from 'effection';
+import { createServer } from 'http';
+import { promisify } from 'util';
+
+import {
+  server as WebSocketServer,
+  request as Request,
+  IMessage as Message
+} from 'websocket';
+
+import { once } from './effection/events';
+import { Mailbox } from './mailbox';
+
+
+interface AgentConnectionServerOptions {
+  port: number;
+  inbox: Mailbox;
+  delegate: Mailbox;
+}
+
+export class AgentConnectionServer {
+  constructor(public options: AgentConnectionServerOptions) {}
+
+  *listen(forEach: (connection: AgentConnection) => Operation): Operation {
+    let { port, inbox, delegate } = this.options;
+
+    let httpServer = createServer();
+    let socketServer = new WebSocketServer({ httpServer });
+
+    let parent = yield ({ resume, context }) => resume(context.parent.parent);
+    parent['spawn'](fork(function*() {
+
+      yield monitor(function* raiseServerErrors() {
+        let [error]: [Error] = yield once(httpServer, "error");
+        throw error;
+      });
+
+      try {
+        yield acceptConnections(socketServer, inbox, delegate, forEach);
+      } finally {
+        socketServer.unmount();
+      }
+    }));
```

Comment:
We should be able to use our new effection primitives here, no?

Reply:
Yes indeed!